### PR TITLE
feat: add import order sorting via prettier

### LINKED ISSUE #1332: feat: Opinionated import order
### Is your feature request related to a problem? Please describe.

Is there any plans on creating an opinionated prettier import auto-sorter for the t3 base template?

### Describe the solution you'd like to see

Something like [@trivago/prettier-plugin-sort-imports](https://www.npmjs.com/package/@trivago/prettier-plugin-sort-imports) or [@ianvs/prettier-plugin-sort-imports](https://www.npmjs.com/package/@ianvs/prettier-plugin-sort-imports) would be very nice to have already set-up.

Code_656t1Kovbi

### Describe alternate solutions

Import order structure example:

```js
importOrder: [
        "^react",
        "^next/*",
        "<THIRD_PARTY_MODULES>",
        "",
        "^@/components/(.*)$",
        "^@/utils/(.*)$",
        "",
        "^@/lib/(.*)$",
        "",
        "^[./]", // local files
        "^@public/*", // root/public
    ],
  ```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by FinnDore**
Closes #1332 

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] I performed a functional test on my final commit

---

## Changelog

Added @ianvs/prettier-plugin-sort-import to the cli, docs and generated cta apps

---

## Screenshots

it also seems to format code blocks which is cool 
image

💯

### CODE REVIEW

**Comment by FinnDore** [comment]
need to actually test this on a generated app

**Comment by FinnDore** [comment]
image
Slight mis calculation, the base cta app does not actually include prettier....

**Comment by iAverages** [comment]
> Slight mis calculation, the base cta app does not actually include prettier....
> we dont just want this to be available if you chose tailwind, however iirc the choice was made not to include prettier by default.


It should be pretty easy to add an option for prettier since it just needs installing + a single config. None of the other files change. I think that would probably be the best option here if installing it by default is a no go.

**Comment by c-ehrlich** [comment]
IMO we should have this either everywhere or nowhere. Limiting it to apps with Tailwind makes no sense. This would obviously mean including Prettier with all apps, but I think it might be worth it. 

So I see three options:
1. Don't implement this
2. Ship a minimal prettier config to everyone
3. Add an option in the CLI that gives an opinionated prettier+eslint config

**Comment by juliusmarminge** [comment]
I think i vote 3

**Comment by FinnDore** [comment]
What would change about the eslint config? The default cta one has served my just fine tbh

**Comment by juliusmarminge** [comment]
> What would change about the eslint config? The default cta one has served my just fine tbh

It would be either on or off i guess?

We can add this too i think https://github.com/t3-oss/create-t3-app/issues/1197

**Comment by FinnDore** [comment]
yeah that sounds good. will have a crack and we can see if its any good. A bit dubious of adding to many options.

**Comment by FinnDore** [comment]
Think its done. Im debating wether it should be `prettierAndStrictEslint` rather than `strictPrettierAndEslint`. or even if there should be two options?

I welcome thaughts.

**Comment by FinnDore** [comment]
amazing. ill see if i can fix this tomorrow. _Rather suffer with eslint config issues than watch the coronation._
```ts
@ct3a/www:lint:
@ct3a/www:lint: Oops! Something went wrong! :(
@ct3a/www:lint:
@ct3a/www:lint: ESLint: 8.36.0
@ct3a/www:lint:
@ct3a/www:lint: RangeError: Error while loading rule 'isaacscript/format-jsdoc-comments': Invalid count value
@ct3a/www:lint: Occurred while linting /Users/finn/Documents/GitHub/create-t3-app/www/dist/_astro/index.2899cbe4.js
@ct3a/www:lint:     at String.repeat (<anonymous>)
@ct3a/www:lint:     at formatText (/Users/finn/Documents/GitHub/create-t3-app/node_modules/.pnpm/eslint-plugin-isaacscript@2.6.7_@typescript-eslint+parser@5.56.0_eslint@8.36.0_typescript@5.0.2/node_modules/eslint-plugin-isaacscript/src/format.js:192:42)
@ct3a/www:lint:     at /Users/finn/Documents/GitHub/create-t3-app/node_modules/.pnpm/eslint-plugin-isaacscript@2.6.7_@typescript-eslint+parser@5.56.0_eslint@8.36.0_typescript@5.0.2/node_modules/eslint-plugin-isaacscript/src/rules/format-jsdoc-comments.js:58:59
@ct3a/www:lint:     at Array.forEach (<anonymous>)
@ct3a/www:lint:     at create (/Users/finn/Documents/GitHub/create-t3-app/node_modules/.pnpm/eslint-plugin-isaacscript@2.6.7_@typescript-eslint+parser@5.56.0_eslint@8.36.0_typescript@5.0.2/node_modules/eslint-plugin-isaacscript/src/rules/format-jsdoc-comments.js:52:23)
@ct3a/www:lint:     at Object.create (/Users/finn/Documents/GitHub/create-t3-app/node_modules/.pnpm/@typescript-eslint+utils@5.56.0_eslint@8.36.0_typescript@5.0.2/node_modules/@typescript-eslint/utils/dist/eslint-utils/RuleCreator.js:41:20)
@ct3a/www:lint:     at createRuleListeners (/Users/finn/Documents/GitHub/create-t3-app/node_modules/.pnpm/eslint@8.36.0/node_modules/eslint/lib/linter/linter.js:926:21)
@ct3a/www:lint:     at /Users/finn/Documents/GitHub/create-t3-app/node_modules/.pnpm/eslint@8.36.0/node_modules/eslint/lib/linter/linter.js:1108:110
@ct3a/www:lint:     at Array.forEach (<anonymous>)
@ct3a/www:lint:     at runRules (/Users/finn/Documents/GitHub/create-t3-app/node_modules/.pnpm/eslint@8.36.0/node_modules/eslint/lib/linter/linter.js:1045:34)
@ct3a/www:lint:  ELIFECYCLE  Command failed with exit code 2.
@ct3a/www:lint: ERROR: command finished with error: command (/Users/finn/Documents/GitHub/create-t3-app/www) pnpm run lint exited (1)
```

**Comment by FinnDore** [comment]
Nvm it was just trying to lint the dist dir

File: cli/.prettierrc.cjs
Comment (FinnDore) [reply cli/.prettierrc.cjs]: good shout. done

**Comment by FinnDore** [comment]
Ive added a format and `format:check` script. Ive also reversed how `check` extends `format` as it was writing when using  `format:check` lmk if this was actually intended, i think i did a similar thing inside www on a previous pr.

File: cli/package.json
Comment (juliusmarminge) [inline cli/package.json:36]: does `--write` override the `--check` in the `:check` script?

bit cursed but i'll allow it

File: cli/src/installers/strictEslintAndPrettier.ts
Comment (juliusmarminge) [inline cli/src/installers/strictEslintAndPrettier.ts:27]: imo we should have all or nothing, so if the installer isn't selected, no linter will be setup?

File: cli/src/installers/tailwind.ts
Comment (juliusmarminge) [inline cli/src/installers/tailwind.ts:32]: we can then remove prettier from here - keep all the prettier logic in the prettier-eslint installer.

File: cli/src/utils/renderVersionWarning.ts
Comment (juliusmarminge) [inline cli/src/utils/renderVersionWarning.ts:11]: i wish import sorters didn't move imports when they are separated by actual code blocks...

is there a setting for this?

File: cli/template/base/package.json
Comment (juliusmarminge) [inline cli/template/base/package.json:20]: this shouldn't be here right? it should be added by the installer?

File: cli/template/base/package.json
Comment (juliusmarminge) [inline cli/template/base/package.json:11]: same for these scripts

File: cli/template/extras/src/server/api/trpc/base.ts
Comment (juliusmarminge) [inline cli/template/extras/src/server/api/trpc/base.ts:28]: same here - wish it didn't move these imports when they are clearly separated :/

we'll have to look over the comments cause those should not be moved

**Review by juliusmarminge — CHANGES_REQUESTED** [review]
Some concerns here.

IMO, a model like this would be better:

base app comes with no eslint n'or prettier.

we add 2 installers - 1 for eslint and 1 for prettier.

if choosing both, we should set up a nice eslint config with the prettier eslint plugin to make sure they work nicely together

if selecting only eslint - nothing changes today, except we remove the prettier part from the tailwind installer

if selecting only prettier - we setup the import sorter, some basic config and the tailwind plugin (if user has selected tailwind ofc)


thoughts?

**Comment by juliusmarminge** [comment]
Also - it might be easier if you separate this pr into 2 to make it easier to review and not being blocked on silly things - one for adding the sorter to the monorepo (should be a quick approval) and one for adding this installer.

File: cli/package.json
Comment (FinnDore) [reply cli/package.json:36]: image
P much

**Comment by FinnDore** [comment]
Yeah i was thinking the same. My only concern was, having a installer that installs a single or a few eslint rules seems a bit weird. If we feel like the rule is benefit the user why dont we include them by default in the already bundled eslint config. Then that brings us back to the "should be be bundling prettier?" question, IMO having a installer for prettier would be great as i ( and im sure others ) always have to run `cp ../some-other-project/.prettier.config.json .` and then install prettier + add format and format lint scripts.

If we were to add a explicit prettier installer removing the prettier stuff from the tailwind installer sounds like a good shout yeah.

I think a base project with no eslint is a non optimal situtation, those web accessibility lints are amazing imo.

thoughts on my thoughts of your thoughts?

File: cli/src/utils/renderVersionWarning.ts
Comment (FinnDore) [reply cli/src/utils/renderVersionWarning.ts:11]: Looks like you can `//prettier-ignore` lines but thats about it :/

**Comment by FinnDore** [comment]
Got a branch for the monorepo changes, the format commands are dying on me though, I'll open the or tomorrow it should be in a ready state by then.

**Comment by FinnDore** [comment]
So we cant just remove eslint from the base project or it wont build... do we want
```
- empty eslint config in the base project + extra rules installed when the eslint option is selected
- just include the new elint rule in the default config and not have a separate installer
```
im also wondering if we should do the eslint portion in a separate pr

**Comment by FinnDore** [comment]
prettier is not formatting some of my files locally aswell smh

File: cli/template/extras/config/_prettier.config.cjs
Comment (DavidRecheni) [inline cli/template/extras/config/_prettier.config.cjs:3]: `imports` *

Comment (juliusmarminge) [inline cli/template/extras/config/_prettier.config.cjs:3]:
```suggestion
  plugins: ["@ianvs/prettier-plugin-sort-imports"],
```

File: cli/template/base/package.json
Comment (FinnDore) [inline cli/template/base/package.json:10]: will change this to only be added when prettier is enabled

Comment (FinnDore) [reply cli/template/base/package.json:10]: ( format and format:check )

File: cli/template/extras/config/_prettier.config.cjs
Comment (FinnDore) [reply cli/template/extras/config/_prettier.config.cjs]: ty btw :)

**Comment by FinnDore** [comment]
Removed the `require.resolve` from the tailwind prettier config as im not sure why it was needed, seems to work, somone lmk if its actually needed

**Comment by FinnDore** [comment]
Tested this all and i think its good to go now. although i thought that 4 days ago so im ready to be proven wrong.

**Comment by juliusmarminge** [comment]
Instead of having this as an installer next to the other pacakges, do you tihnk it'd be better to have like

```bash
? What pacakges would you like...
  o nextAuth
  ...

? Would you like to enable a more opinionated eslint+prettier setup?
  x yes o no
```

**Comment by FinnDore** [comment]
Good shout, maybe the wording should different though as it sounds like a more strict prettier config is added aswell which is wrong

**Comment by FinnDore** [comment]
I should probably update the ci aswell....

**Comment by FinnDore** [comment]
cant tell if ive borked the e2e workflow file or if it need approval to run since ive changes it

**Comment by juliusmarminge** [comment]
Merging to intermediate branch to release a canary

**Comment by juliusmarminge** [comment]
bruh - based the intermediate branch off the wrong thing...

can you reopen 😄

**Pull request merged**
